### PR TITLE
fix #1272 subresource recursive + 1 more test

### LINKED ISSUE #1272: Subresource visited tracking is broken
Basically the same as #1139. Previously fixed in #1144 and #1162.

> Category
> - children
> - parent

Self-referencing resources results in infinite recursion leading to OOM.

> Product
> - brands
> - isRelatedTo
> - isSimilarTo

It tries to traverse deeper on the same path more than once, e.g.

> [LogicException]
> Route pattern "/products/{id}/is_related_to/{isRelatedTo}/is_related_to/{isRelatedTo}/brands.{_format}"
> cannot reference variable name "isRelatedTo" more than once.

The expected behaviour is:
```
Product -[isRelatedTo]-> Product -[isRelatedTo] -> Product
                                 /  cut here  /
```

### PR DESCRIPTION
**Pull request by soyuka**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #1272
| License       | MIT
| Doc PR        | api-platform/doc#1234

### CODE REVIEW

**Comment by soyuka** [comment]
Whew, finally done. Wait till tests pass then merge ;).

Helped discover a bug in swagger (duplicate properties (filter + subresource identifier)) and duplicate tags (in case of self-referencing).

File: src/Operation/Factory/SubresourceOperationFactory.php [hunk 1]
```diff
@@ -79,18 +79,13 @@ private function computeSubresourceOperations(string $resourceClass, array &$tre
...
-                $prefix = '';
-                $visiting = "{$parentOperation['property']}-{$parentOperation['resource_class']}-$property-$subresourceClass";
+                $suffix = "$property-$subresourceClass";
 
-                foreach ($parentOperation['identifiers'] as $key => list($param, $class)) {
-                    $prefix .= 0 === $key ? $class : "-$param-$class";
-                }
-
-                if (false !== strpos($prefix, $visiting)) {
+                if (false !== strpos($parentVisiting, $suffix)) {
```
Comment (teohhanhui) [inline src/Operation/Factory/SubresourceOperationFactory.php]: We're still doing `strpos`? Is there any downside to the algorithm I've described? It's much easier to understand, at least...

**Comment by teohhanhui** [comment]
I think this was merged too soon. I'd have loved to review it...

**Comment by curry684** [comment]
I ran into this bug 4 times past week so I for one appreciate the quick fix. It's always possible to have a second PR iron out the issues on master  ;)

**Comment by teohhanhui** [comment]
Actually, I agree. @soyuka is awesome!

**Comment by Simperfit** [comment]
I agree with @teohhanhui ;)

**Comment by soyuka** [comment]
:laughing: :+1: @curry684 !

@teohhanhui I wanted to quick fix this before the beta release and that @dunglas went on vacation! Feel free to test and report back, we can always improve this :).

**Pull request merged**